cabal-validate/main/Main.hs

```diff
@@ -0,0 +1,953 @@
+import Control.Exception (Exception (displayException), catch, throw, throwIO)
+import Control.Monad (forM_, unless)
+import Data.Data (Typeable)
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Maybe (listToMaybe)
+import Data.Text (Text)
+import qualified Data.Text as T
+import Data.Time.Clock (DiffTime, secondsToDiffTime)
+import Data.Time.Clock.System (getSystemTime, systemToTAITime)
+import Data.Time.Clock.TAI (AbsoluteTime, diffAbsoluteTime)
+import Data.Time.Format (defaultTimeLocale, formatTime)
+import Data.Version (Version, makeVersion, parseVersion, showVersion)
+import GHC.Conc (getNumCapabilities)
+import Options.Applicative
+  ( FlagFields
+  , Mod
+  , Parser
+  , ParserInfo
+  , auto
+  , execParser
+  , flag
+  , flag'
+  , fullDesc
+  , help
+  , helper
+  , hidden
+  , info
+  , long
+  , maybeReader
+  , option
+  , progDesc
+  , short
+  , strOption
+  , switch
+  , value
+  )
+import qualified Options.Applicative as Opt
+import System.Console.ANSI
+  ( Color (Blue, Cyan, Green, Red)
+  , ColorIntensity (Vivid)
+  , ConsoleIntensity (BoldIntensity)
+  , ConsoleLayer (Foreground)
+  , SGR (Reset, SetColor, SetConsoleIntensity)
+  , setSGRCode
+  )
+import System.Console.Terminal.Size (Window (width), size)
+import System.Directory (getCurrentDirectory)
+import System.Exit (ExitCode (ExitFailure, ExitSuccess), exitFailure, exitSuccess)
+import System.Info (arch, fullCompilerVersion, os)
+import Text.ParserCombinators.ReadP (readP_to_S)
+import Turtle (ProcFailed (ProcFailed, procArguments, procCommand, procExitCode), cd, proc, procStrict, when, (</>))
+
+tPutStrLn :: Text -> IO ()
+tPutStrLn = putStrLn . T.unpack
+
+tShow :: Show a => a -> Text
+tShow = T.pack . show
+
+tSetSGRCode :: [SGR] -> Text
+tSetSGRCode = T.pack . setSGRCode
+
+-- | Command-line options, resolved with context from the environment.
+data ResolvedOpts = ResolvedOpts
+  { verbose :: Bool
+  , jobs :: Int
+  , cwd :: FilePath
+  , startTime :: AbsoluteTime
+  , compiler :: Compiler
+  , extraCompilers :: [FilePath]
+  , cabal :: FilePath
+  , hackageTests :: HackageTests
+  , archPath :: FilePath
+  , projectFile :: FilePath
+  , tastyArgs :: [Text]
+  , targets :: [Text]
+  , steps :: [Step]
+  }
+  deriving (Show)
+
+data Compiler = Compiler
+  { compilerExecutable :: FilePath
+  , compilerVersion :: Version
+  }
+  deriving (Show)
+
+data VersionParseException = VersionParseException
+  { versionInput :: String
+  , versionExecutable :: FilePath
+  }
+  deriving (Typeable, Show)
+
+instance Exception VersionParseException where
+  displayException exception =
+    "Failed to parse `"
+      <> versionExecutable exception
+      <> " --numeric-version` output: "
+      <> show (versionInput exception)
+
+makeCompiler :: FilePath -> IO Compiler
+makeCompiler executable = do
+  -- TODO: Check the exit code!
+  (_exitCode, stdout) <- procStrict (T.pack executable) ["--numeric-version"] ""
```

Comment:
Is it `turtle` forcing us to convert between `Text` and `String` everywhere? Is there enough added value in comparison to `process`?

Reply:
Done, switched over to `typed-process` and the code is a lot nicer now. It's a shame that `text` isn't the standard though, UTF-8 is really so much better than a linked list of 4-byte code units.